### PR TITLE
Cassandra multinode test for Atlas working with Cassandra HA

### LINKED ISSUE #819: AtlasDB starts with single node failure of Cassandra
We should have automated test coverage for AtlasDB clients attempting to connect to a 3-node Cassandra cluster, and one node is unavailable. AtlasDB should be able to startup and gracefully handle a Cassandra node failure.

We haven't observed any issues with this in practice, but it is not covered by automated testing at the moment.

### PR DESCRIPTION
**Pull request by hsaraogi**
**Update 27/9 by Glenn:**
We now have passing tests (modulo flickers)!
I would like someone to look over the CassandraMultinodeEteTest file once again now that I've finished working on it.
In particular, different code paths will work depending on whether the test is running locally or on circle. I've done this in a fairly hacky way (try the circle way, then the local way) - any suggestions/requests for improvement?

First part of fix for #819.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/927)

### CODE REVIEW

File: atlasdb-ete-tests/docker-compose.multiple-cassandra.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+version: '2'
+
+services:
+  cassandra1:
+     image: com.palantir.docker/atlas-cassandra:2.2-v0.1
+     environment:
+      - CASSANDRA_SEEDS=cassandra1,cassandra2
```
Comment (schlosna) [inline atlasdb-ete-tests/docker-compose.multiple-cassandra.yml]: Should we include cassandra3 as a seed or are we testing the cluster gossiping correctly to pick it up?

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeTestSuite.java [hunk 2]
```diff
@@ -18,6 +18,7 @@
 import java.util.List;
 
 import org.junit.ClassRule;
+import org.junit.Ignore;
```
Comment (schlosna) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeTestSuite.java]: Is this unused?

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraNoLeaderTestSuite.java [hunk 3]
```diff
@@ -18,6 +18,7 @@
 import java.util.List;
 
 import org.junit.ClassRule;
+import org.junit.Ignore;
```
Comment (schlosna) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraNoLeaderTestSuite.java]: Is this used?

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/DbKvsTestSuite.java [hunk 4]
```diff
@@ -18,6 +18,7 @@
 import java.util.List;
 
 import org.junit.ClassRule;
+import org.junit.Ignore;
```
Comment (schlosna) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/DbKvsTestSuite.java]: Unused?

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeTodoEteTest.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    @Test
+    public void shouldRunTransactionsAfterCassandraNodeIsShutDownWithoutUnacceptableDelay()
+            throws InterruptedException {
+        TodoResource clientToSingleNode = EteSetup.createClientToSingleNode(TodoResource.class);
+
+        assertAddTodoTransactionWasSuccessful(clientToSingleNode);
+
+        String cassandraNodeToKill = getRandomCassandraNodeToShutdown();
```
Comment (schlosna) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeTodoEteTest.java]: How did we pick these timings? We should be able to do a quorum write with 2 or 3 nodes available in roughly the same time range assuming we talk to a live node that is not partitioned away from the cluster. If we try talking to a dead or partitioned node, I could see needing to wait for timeout & retry. 

With the random based Russian roulette here, we get different behavior each time ETE runs -- a 1 in 3 chance the client we talk to is the one killed. We might want to expand the tests to cover these scenarios more explicitly.

File: scripts/circle-ci/run-circle-tests.sh [hunk 6]
```diff
@@ -31,7 +31,7 @@ done
 case $CIRCLE_NODE_INDEX in
     0) ./gradlew --profile --continue check $CONTAINER_0_EXCLUDE_ARGS ;;
     1) ./gradlew --profile --continue --parallel ${CONTAINER_1[@]} ;;
-    2) ./gradlew --profile --continue --parallel ${CONTAINER_2[@]} ;;
+    2) ./gradlew --profile --continue ${CONTAINER_2[@]} ;;
```
Comment (schlosna) [inline scripts/circle-ci/run-circle-tests.sh]: Might want to comment why this isn't/ can't be parallelized

File: atlasdb-ete-tests/docker-compose.multiple-cassandra.yml (hunk 1 above)
Comment (hsaraogi) [inline atlasdb-ete-tests/docker-compose.multiple-cassandra.yml]: We don't need 3 seeds as only one node is supposed to be down at a time, so when it comes up, it should still be able to become aware of all three by contacting to the seed node that was up.

**Comment by tpetracca** [comment]
@ashrayjain for SA on some ETE test stuff

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            .waitingForService("cassandra1", Container::areAllPortsOpen)
+            .waitingForService("cassandra2", Container::areAllPortsOpen)
+            .waitingForService("cassandra3", Container::areAllPortsOpen)
+            .waitingForService("ete1", toBeReady())
+            .saveLogsTo(CONTAINER_LOGS_DIRECTORY)
+            .build();
+
+
+    @ClassRule
+    public static final RuleChain CASSANDRA_DOCKER_SET_UP = RuleChain
```
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Differentiating between `SET_UP` and `SETUP` is really nasty. Perhaps this one can be `PREPARED_DOCKER_SETUP` and the inner one can be `MULTINODE_CASSANDRA_SETUP`.

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        String cassandraNodeToKill = getRandomCassandraNodeToShutdown();
+        killCassandraContainer(cassandraNodeToKill);
+
+        long transactionStartTime = System.currentTimeMillis();
+        assertAddTodoTransactionWasSuccessful(clientToSingleNode);
+        long transactionEndTime = System.currentTimeMillis();
+
+        long transactionTimeAfterNodeIsKilled = transactionEndTime - transactionStartTime;
+
+        startCassandraContainer(cassandraNodeToKill);
```
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: If this is the end of the test, do we need to restart the container at all?

Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: I guess a fuller ETE test would be:

- Gets work with all nodes up
- Deletes work with all nodes up
- Shut down a node
- Gets work with one node down
- Deletes fail with one node down (and RF3)
- Start up the node
- Deletes work again

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    private void assertAddTodoTransactionWasSuccessful(TodoResource todoClient) {
+        Todo todo = ImmutableTodo.of("some unique TODO item with UUID=" + UUID.randomUUID());
+
+        todoClient.addTodo(todo);
+        List<Todo> todoList = todoClient.getTodoList();
+
+        assertThat(todoList, hasItem(todo));
+    }
+
+    private void waitForNodetoolToConfirmStatus(Container container, String status, int expectedNodeCount) {
```
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Sanity check (as I've heard of a problem bringing the container back to life): does this work in the happy case where all nodes stay up?

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java (hunk 8 above)
Comment (hsaraogi) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: @gsheasby 
More cases:
- Gets work when the node to which the query is directed to is down (with large transaction time)
- Deletes work when the node to which the query is directed to is down (with large transaction time)
- Gets work when a node to which the query is not directed to is down (with small transaction time)
- Deletes work when a node to which the query is not directed to is down (with small transaction time)

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java (hunk 9 above)
Comment (hsaraogi) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: If we separate out the cases for which node the query is directed or not, then we should find a way to see that the node we just started joined Cassandra.

Docker-exec doesnt run on Circle, this works fine locally in all cases.

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java (hunk 8 above)
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: It's worth considering doing these incrementally.

**Comment by gsheasby** [comment]
Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java, line 74 at r7](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KRQL5JMrAinoMnTNGUq-r7-74:-KSewatsvWUVyw7lOnxH:b-x7l0v) ([raw file](https://github.com/palantir/atlasdb/blob/ff017317c2dbeabec9340268672efc5240af0584/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java#L74)):*
<details><summary><i>Previously, schlosna (David Schlosnagle) wrote…</i></summary>
> How did we pick these timings? We should be able to do a quorum write with 2 or 3 nodes available in roughly the same time range assuming we talk to a live node that is not partitioned away from the cluster. If we try talking to a dead or partitioned node, I could see needing to wait for timeout & retry. 
> 
> 
> 
> With the random based Russian roulette here, we get different behavior each time ETE runs -- a 1 in 3 chance the client we talk to is the one killed. We might want to expand the tests to cover these scenarios more explicitly.
</details>
I imagine these timings were empirical - @hsaraogi?

I don't think we can force ourselves to talk to an up (or indeed the down) C* node when we add/get the todos. It's possibly an interesting extension to do the call three times and measure the perf of each one, and I'm open to doing that in a future PR, but for now (in spite of the "unacceptable delay" promise in the title) it's useful to know that the call works at all.

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java, line 123 at r9](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KRclBWolCHMZnMt0CCZ-r9-123:-KSex8Y1Vm14HWTfL83Z:be3jr9c) ([raw file](https://github.com/palantir/atlasdb/blob/2831111e9eadc02bc00c77f4a1d7397b7ed6fbc0/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java#L123)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> It's worth considering doing these incrementally.
</details>
Right now the test does:
* Puts & Gets work with all nodes up
* Shut down a node
* Puts & Gets work with one node down
* Start up the node
* Node joins the cluster correctly

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java, line 221 at r9](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KRclBWolCHMZnMt0CCZ-r9-221:-KSexVlmC0NRiu5q2jtY:b-10030x) ([raw file](https://github.com/palantir/atlasdb/blob/2831111e9eadc02bc00c77f4a1d7397b7ed6fbc0/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java#L221)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary>
> If we separate out the cases for which node the query is directed or not, then we should find a way to see that the node we just started joined Cassandra.
> 
> 
> 
> Docker-exec doesnt run on Circle, this works fine locally in all cases.
</details>
Now works both locally and circle-y.

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeTestSuite.java, line 21 at r7](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KRQL5JMrAinoMnTNGUm-r7-21:-KSexc11mEKRafhO7irZ:b-nbmdot) ([raw file](https://github.com/palantir/atlasdb/blob/ff017317c2dbeabec9340268672efc5240af0584/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeTestSuite.java#L21)):*
<details><summary><i>Previously, schlosna (David Schlosnagle) wrote…</i></summary>
> Is this unused?
</details>
Not any more.

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraNoLeaderTestSuite.java, line 21 at r7](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KRQL5JMrAinoMnTNGUn-r7-21:-KSexh2QGf0kYyU6cOmw:b-pvphg1) ([raw file](https://github.com/palantir/atlasdb/blob/ff017317c2dbeabec9340268672efc5240af0584/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraNoLeaderTestSuite.java#L21)):*
<details><summary><i>Previously, schlosna (David Schlosnagle) wrote…</i></summary>
> Is this used?
</details>
Yes, it is now.

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/DbKvsTestSuite.java, line 21 at r7](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KRQL5JMrAinoMnTNGUo-r7-21:-KSexjJGQOpQLRHuhEJa:b-mfsog3) ([raw file](https://github.com/palantir/atlasdb/blob/ff017317c2dbeabec9340268672efc5240af0584/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/DbKvsTestSuite.java#L21)):*
<details><summary><i>Previously, schlosna (David Schlosnagle) wrote…</i></summary>
> Unused?
</details>
Not any more. By the way, checkstyle does this check for us :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/927)*

**Comment by gsheasby** [comment]
Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[scripts/circle-ci/run-circle-tests.sh, line 34 at r7](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KREQ8am3yEyVHVFSaZW-r7-34:-KSeyjfDahavQx43cghi:b-huj9a) ([raw file](https://github.com/palantir/atlasdb/blob/ff017317c2dbeabec9340268672efc5240af0584/scripts/circle-ci/run-circle-tests.sh#L34)):*
<details><summary><i>Previously, schlosna (David Schlosnagle) wrote…</i></summary>
> Might want to comment why this isn't/ can't be parallelized
</details>
Outside the scope of this PR. @jboreiko wrote the non-parallel line 2 months ago.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/927)*

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+    @Test
+    public void shouldRunTransactionsWithAllCassandraNodesRunningWithoutUnacceptableDelay()
+            throws InterruptedException, IOException {
+        TodoResource clientToSingleNode = createClientFor(TodoResource.class, asPort("ete1"));
+
+        long transactionStartTime = System.currentTimeMillis();
+        assertAddTodoTransactionWasSuccessful(clientToSingleNode);
+        long transactionEndTime = System.currentTimeMillis();
+
+        long transactionTimeWithAllNodesRunning = transactionEndTime - transactionStartTime;
```
Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Can we extract this duplicated timing code? Maybe a utility function like 

```
    private long timeit(Runnable task) {
        long transactionStartTime = System.currentTimeMillis();
        task.run();
        long transactionEndTime = System.currentTimeMillis();
        return transactionEndTime - transactionStartTime;
    }
```

and then use
```
long transactionTimeAfterNodeIsKilled = timeit(() -> assertAddTodoTransactionWasSuccessful(clientToSingleNode));
```

Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Actually, if we only ever time `assertAddTodoTransactionWasSuccessful(...)`, just extract `timeSuccessfulAddTodoTransaction(...)`

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+    private static final Gradle GRADLE_PREPARE_TASK =
+            Gradle.ensureTaskHasRun(":atlasdb-ete-test-utils:buildCassandraImage");
+
+    private static final int TIMELOCK_SERVER_PORT = 3828;
+
+    private static final String CONTAINER_LOGS_DIRECTORY = "container-logs/cassandra-multinode";
+
+    private static final List<String> CASSANDRA_NODES = ImmutableList.of("cassandra1", "cassandra2", "cassandra3");
+
+    private static final long MAX_CASSANDRA_NODE_DOWN_MILLIS = 30000;
```
Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Nit: `30_000` is easier to read

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+        long transactionEndTime = System.currentTimeMillis();
+
+        long transactionTimeWithAllNodesRunning = transactionEndTime - transactionStartTime;
+
+        assertThat("transactionTimeWithAllNodesRunning",
+                transactionTimeWithAllNodesRunning,
+                is(lessThan(MAX_CASSANDRA_NODES_RUNNING_MILLIS)));
+
+        String container = CASSANDRA_NODES.get(0);
+        checkNodetoolStatus(MULTINODE_CASSANDRA_SETUP.containers().container(container), "UN", 3);
```
Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Is this check necessary?

Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Maybe a safer/more general way to do this is via a `@Before` check to restart any previously killed containers.

File: atlasdb-ete-tests/docker-compose.multiple-cassandra.yml [hunk 13]
```diff
@@ -0,0 +1,37 @@
+version: '2'
```
Comment (gbrova) [inline atlasdb-ete-tests/docker-compose.multiple-cassandra.yml]: Kind of weird for these docker-compose files to live in the root directory. Consider moving them to docker/compose or similar.

(I'm happy for this to happen in a separate PR)

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    @Test
+    public void shouldRunTransactionsAfterCassandraNodeIsShutDownWithoutUnacceptableDelay()
```
Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Nit: Not a fan of the double-negative "WithoutUnacceptable". Consider: `shouldRunTransactionFastEnoughWithOneCassandraNodeDown`

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+            .build();
+
+
+    @ClassRule
+    public static final RuleChain PREPARED_DOCKER_SETUP = RuleChain
+            .outerRule(GRADLE_PREPARE_TASK)
+            .around(MULTINODE_CASSANDRA_SETUP);
+
+    @Test
+    public void shouldRunTransactionsWithAllCassandraNodesRunningWithoutUnacceptableDelay()
```
Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Nit: Not a fan of the double-negative "WithoutUnacceptable". Consider: `shouldRunTransactionFastEnoughWithAllCassandraNodesUp`

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+            });
+        };
+    }
+
+    private String getRandomCassandraNodeToShutdown() {
+        int index = ThreadLocalRandom.current().nextInt(CASSANDRA_NODES.size());
+        return CASSANDRA_NODES.get(index);
+    }
+
+    private static void killCassandraContainer(String containerName) {
```
Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Is it worth moving all the cassandra node orchestration stuff below into a separate class, to separate the orchestration logic from our tests?

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+        waitForNodetoolToConfirmStatus(container, "UN", 3);
+    }
+
+    private static void waitForAllPorts(Container container) {
+        Awaitility.await()
+                .atMost(50, TimeUnit.SECONDS)
+                .until(() -> container.areAllPortsOpen().succeeded());
+    }
+
+    private void assertAddTodoTransactionWasSuccessful(TodoResource todoClient) {
```
Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: This duplicates `TodoEteTest.shouldBeAbleToWriteAndListTodos()`

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java [hunk 18]
```diff
@@ -0,0 +1,228 @@
+/**
```
Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Are you sure this test gets run? `build.gradle` is set up to only run TestSuites, and I don't see this added as part of a suite.

Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Consider renaming it something like `CassandraMultinodePerformanceEteTest` and adding it to `CassandraMultinodeTestSuite`

Comment (gbrova) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Actually, don't do that. I got confused because "CassandraMultinode" means both of:

* "multiple Atlas nodes using the Cassandra KVS" (in CassandraMultinodeTestSuite)
* "multiple Cassandra nodes and one ETE node" (in this file, CassandraMultinodeEteTest)

Instead, we probably want to create a new TestSuite that represents this setup, and that contains these extra tests.

Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: It has *definitely* failed on circle.

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java (hunk 12 above)
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Good catch. I'd left that line in while testing the giraffe stuff.

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java (hunk 17 above)
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraMultinodeEteTest.java]: Proposed solution: extract the logic into something like "TodoEteTestUtils". What do you think?

File: atlasdb-ete-tests/docker-compose.multiple-cassandra.yml (hunk 13 above)
Comment (gsheasby) [inline atlasdb-ete-tests/docker-compose.multiple-cassandra.yml]: Definitely to go in a separate PR. Would be useful to keep a list somewhere (a github issue) with the Cassandra HA outstanding things, so we don't lose sight of them.

**Comment by gsheasby** [comment]
Closing due to other priorities and the failing test. I will work on fixing the test on the branch temp/make-ha-test-run: https://github.com/palantir/atlasdb/tree/temp/make-ha-test-run - and will reopen this PR once I've fixed the test.

**Comment by gsheasby** [comment]
Reopening because I'm scared of the delete button and we *do* definitely want to merge this eventually.

**Comment by jboreiko** [comment]
Reviewed 1 of 9 files at r4, 1 of 18 files at r7, 4 of 7 files at r8, 1 of 2 files at r10.
Review status: 6 of 10 files reviewed at latest revision, 21 unresolved discussions, some commit checks broke.

---

*[atlasdb-ete-tests/build.gradle, line 25 at r12](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KTFRAkEkNCksZvbHVpE:-KTFRAkEkNCksZvbHVpF:b-yul7u0) ([raw file](https://github.com/palantir/atlasdb/blob/9b464adaa3316fd7f521c8164b4196e76238aea4/atlasdb-ete-tests/build.gradle#L25)):*
> ```Gradle
>     compile 'com.google.dagger:dagger:' + libVersions.dagger
> 
>     compile('com.palantir.giraffe:giraffe-core:0.8.0')
> ```

These don't appear needed at compile

---

*[atlasdb-ete-tests/docker-compose.multiple-cassandra.yml, line 7 at r7](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KREQ8aiwgHnN_NJL8sH-r7-7:-KTFRRmRtjuzpQObyslZ:bnysh2j) ([raw file](https://github.com/palantir/atlasdb/blob/ff017317c2dbeabec9340268672efc5240af0584/atlasdb-ete-tests/docker-compose.multiple-cassandra.yml#L7)):*
<details><summary><i>Previously, hsaraogi (Himangi Saraogi) wrote…</i></summary>
> We don't need 3 seeds as only one node is supposed to be down at a time, so when it comes up, it should still be able to become aware of all three by contacting to the seed node that was up.
</details>
It does seem a bit odd to have this since I believe in most deployment envs all nodes are seed nodes.

---

*[atlasdb-ete-tests/docker/conf/atlasdb-ete.multiple-cassandra.yml, line 2 at r12](https://reviewable.io:443/reviews/palantir/atlasdb/927#-KTFRnz213Zsym9ZSfVA:-KTFRnz213Zsym9ZSfVB:blzdnmg) ([raw file](https://github.com/palantir/atlasdb/blob/9b464adaa3316fd7f521c8164b4196e76238aea4/atlasdb-ete-tests/docker/conf/atlasdb-ete.multiple-cassandra.yml#L2)):*
> ```YAML
> server:
>   #rootPath: /api/*
> ```

Why do we have defaults commented out in here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/927)*

**Comment by harrybiddle** [comment]
I have removed the nodetool check that was causing the CircleCI failure and have brought the PR up to date with develop. The CircleCI tests are now passing, modulo test flickers.

However, I think this PR _introduces_ a test flicker. Assigning to @SerialVelocity to fix the flicker and remove some tech debt.

@gsheasby / @jboreiko - shout if there is anything else outstanding from the above code reviews.
